rust/kernel/xarray.rs

```diff
@@ -0,0 +1,296 @@
+    /// Creates a new `XArray` with the given flags.
+    pub fn new(flags: Flags) -> Result<XArray<T>> {
+        let xa = Opaque::uninit();
+
+        // SAFETY: We have just created `xa`. This data structure does not require
+        // pinning.
+        unsafe { bindings::xa_init_flags(xa.get(), flags) };
+
+        // INVARIANT: Initialize the `XArray` with a valid `xa`.
+        Ok(XArray {
+            xa,
+            _p: PhantomData,
+        })
+    }
+
+    /// Replaces an entry with a new value, returning the old value (if any).
+    pub fn replace(&self, index: usize, value: T) -> Result<Option<T>> {
+        let new = value.into_foreign();
+
+        let old = unsafe {
+            bindings::xa_store(
+                self.xa.get(),
+                index.try_into()?,
+                new as *mut _,
+                bindings::GFP_KERNEL,
+            )
+        };
+
+        let err = unsafe { bindings::xa_err(old) };
+        if err != 0 {
+            // Make sure to drop the value we failed to store
+            unsafe { T::from_foreign(new) };
+            Err(Error::from_kernel_errno(err))
+        } else if old.is_null() {
+            Ok(None)
+        } else {
+            Ok(Some(unsafe { T::from_foreign(old) }))
+        }
+    }
+
+    /// Replaces an entry with a new value, dropping the old value (if any).
+    pub fn set(&self, index: usize, value: T) -> Result {
+        self.replace(index, value)?;
+        Ok(())
+    }
+
+    /// Looks up and returns a reference to an entry in the array, returning a `Guard` if it
+    /// exists.
+    ///
+    /// This guard blocks all other actions on the `XArray`. Callers are expected to drop the
+    /// `Guard` eagerly to avoid blocking other users, such as by taking a clone of the value.
+    pub fn get(&self, index: usize) -> Option<Guard<'_, T>> {
+        // SAFETY: `self.xa` is always valid by the type invariant.
+        let p = unsafe {
+            bindings::xa_lock(self.xa.get());
+            bindings::xa_load(self.xa.get(), index.try_into().ok()?)
+        };
+
+        if p.is_null() {
+            unsafe { bindings::xa_unlock(self.xa.get()) };
+            None
+        } else {
+            Some(Guard(p as _, &self.xa))
+        }
+    }
+
+    /// Removes and returns an entry, returning it if it existed.
+    pub fn remove(&self, index: usize) -> Option<T> {
+        let p = unsafe { bindings::xa_erase(self.xa.get(), index.try_into().ok()?) };
+        if p.is_null() {
+            None
+        } else {
+            Some(unsafe { T::from_foreign(p) })
+        }
+    }
+
+    /// Allocate a new index in the array, optionally storing a new value into it, with
+    /// configurable bounds for the index range to allocate from.
+    ///
+    /// If `value` is `None`, then the index is reserved from further allocation but remains
+    /// free for storing a value into it.
+    pub fn alloc_limits(&self, value: Option<T>, min: u32, max: u32) -> Result<usize> {
```

Comment:
Given we already have the `reserve_*` APIs, maybe better not providing another way for user to reserve slot?

How about a private `__alloc_limits`? And `alloc_*` APIs look like the below
```rust
fn __alloc_limits(&self, value: Option<T>, min: u32, max: u32) -> Result<usize>

pub fn alloc_limits(&self, value: T, min: u32, max: u32) -> Result<usize>
pub fn alloc(&self, value: T, min: u32, max: u32) -> Result<usize>
```

Reply:
Done for the upstream submission, thanks!